feat: add `srcset` support (`<img>`)

ushu opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

`srcset` attribute is not loaded from the HTML content


**What is the new behavior?**

`srcset` attribute (actually the images inside the `srcset` declaration) is loaded

**Does this PR introduce a breaking change?**
- [ ] Yes
- [x] No

**Other information**:

I did this code *before* realising that an old PR (#55) has been addressing the same issue.
So the code ended up being quite different.

The added value of this version is that it supports multiple qualifiers for the same source such as "100w 2x", which are useful when coding for the mobile web.

Also I am not why this feature has been removed but since I have some time his week I can iterate on the PR with some guidance.

Regards !

cupojoe commented:
What is the status of this making it to release?

michael-ciniawsky commented on lib/attributesParser.js:
```diff
@@ -4,13 +4,36 @@
 */
 var Parser = require("fastparse");
 
+var srcsetQualifierRegexp = /(\s+(\d+w))?(\s+(\d+x))?\s*$/;
```
`var SRCSET_REGEXP`

michael-ciniawsky commented on lib/attributesParser.js:
```diff
@@ -4,13 +4,36 @@
+
+	// allow to load several urls in srcsets, separated by commas
+	var start = index + strUntilValue.length;
+	var subMatches = value.split(/,/);
```
`\n`

michael-ciniawsky commented on lib/attributesParser.js:
```diff
@@ -4,13 +4,36 @@
+		var length = subMatch.length;
+
+		// remove initial spacing
+		var initialSpace = /^\s+/.exec(subMatch);
```
```diff
- var initialSpace = /^\s+/.exec(subMatch);
- var spaceLength = initialSpace ? initialSpace[0].length : 0;
```

```js
var space = {
  initial: /^\s+/.exec(subMatch),
  length: this.initial ? this.initial[0].length : 0
}
```

michael-ciniawsky commented on lib/attributesParser.js:
```diff
@@ -4,13 +4,36 @@
+		var spaceLength = initialSpace ? initialSpace[0].length : 0;
+
+		// remove srcset qualifiers (2x, 110w, etc.), if any
+		var qualifier = srcsetQualifierRegexp.exec(subMatch);
```
`\n`

michael-ciniawsky commented on lib/attributesParser.js:
```diff
@@ -4,13 +4,36 @@
 */
```
Please move this into its own folder && file => `./lib/attrs/srcset.js` and `require('./attrs/srcset.js')` it

michael-ciniawsky commented:
@cupojoe It's in review and needs approval from 2+ maintainers, so no release date estimate can be made yet

d3viant0ne commented:
For reference, one of those should be @hemanth - This is one of the loaders that actually has an assigned lead maintainer

ushu commented:
Hey there,

I moved the code into attrs/srcset.js` as requested, and also updated naming and general structure following your guidance to make the code more readable.

@michael-ciniawsky about:

> What happens if one of these asset urls doesn't exist ?
This maybe better implemented as an opt-in ?

I'm not really sure... the idea in the code was really to load images from the `srcset` as if they were included in a `src` attribute and so I don't think it should accept mentions to non-existing resources.

I know of other attempts to auto-generate these responsive assets on-the-fly, but in my case I work with designers that produce these images (sometimes slightly modified for smaller res) and it is helpful to have `webpack` check if they are present, as it does for `src`.

just my two cents though,
Kind Regards

cupojoe commented:
@michael-ciniawsky @ushu 
Totally agree with @ushu. Assuming the build can generate the images is not a good idea. It should respect the existing sources to be more friendly towards integration with design.

michael-ciniawsky commented:
Sry but I have to declined this since `v1.0.0` of the loader will be a complete overhaul. You can use [`posthtml-loader`](https://github.com/posthtml/posthtml-loader) && write your own/a plugin to achieve this
